File: Low Level Design (LLD)/Problems/05. FoodDeliverySystem/Java/service/CouponService.java
package service;

import payment.Coupon;

import java.util.HashMap;
import java.util.Map;

public class CouponService {
    private Map<String, Coupon> coupons = new HashMap<>();

    public void addCoupon(Coupon coupon) {
        coupons.put(coupon.getCode(), coupon);
    }

    public Coupon getCouponByCode(String code) {
        return coupons.get(code);
    }
}

File: Low Level Design (LLD)/Problems/05. FoodDeliverySystem/Cpp/service/src/CouponService.cpp
#include "../include/CouponService.h"

void CouponService::addCoupon(const Coupon& coupon) {
    coupons[coupon.getCode()] = coupon;
}

const Coupon* CouponService::getCouponByCode(std::string& code) const  {
    auto it = coupons.find(code);
    if (it != coupons.end()) {
        return &it->second;
    }
    return nullptr;
}